Permission policy modules

gasman opened the pull request:
This PR implements permission policies, as originally proposed in https://github.com/torchbox/wagtail/issues/1384#issuecomment-142655487. These are objects that provide a central point for making permission queries for a particular model - such as "can user X perform action Y on instance Z?" or "which instances can user X perform action Y on?". By collecting this logic into a single module, rather than having it distributed across templates, views and models, it's possible to redefine the permission rules for an app by swapping in a new policy object.

The currently defined permission policies are:

* `BlanketPermissionPolicy` - allow everyone (including anonymous users) to do everything, effectively disabling permission rules entirely
* `AuthenticationOnlyPermissionPolicy` - allow all active, authenticated users to do everything
* `ModelPermissionPolicy` - apply standard `django.contrib.auth` permission rules, which act on the model as a whole (i.e. it's not possible to give a user edit access to a subset of a model's instances)
* `OwnershipPermissionPolicy` - the current behaviour of documents / images, appropriate for models that have a concept of 'ownership', where a user with 'add' permission automatically receives permission to change/delete instances they own.

wagtailimages, wagtaildocs and wagtailredirects have been converted to use permission policies, along with wagtailadmin's generic views (which means that wagtailsites uses them as well).

The immediate use-case for this will be to switch wagtailimages and wagtaildocs to the collections system (#1751), but with further work it will be possible to:

* Allow individual Wagtail installations to customise permission policies via settings (#1384). I haven't worked out the best way of doing this - could we pass it as an appconf parameter in INSTALLED_APPS, perhaps?
* Customise the page permission model (#1993). I haven't attempted this here as page permissions will be significantly more complex - more operations to consider (e.g. move / copy), more business rules to account for (e.g. parent_page_types / subpage_types), and the UserPermissionProxy / PagePermissionTester pattern (which would be a useful extension to the permission policy API in general, as it avoids repeated DB hits on listing pages).
* Build up the group permission interface programmatically so that it outputs just the checkboxes relevant to the permission policy in use, rather than the app having to specifically register them. (For example - BlanketPermissionPolicy and AuthenticationOnlyPermissionPolicy would not add anything to the permissions interface, as there's nothing to configure; ModelPermissionPolicy would supply checkboxes for add, change and delete, and OwnershipPermissionPolicy would supply checkboxes for add and change, as it doesn't distinguish between change and delete permissions.)
* Potentially, implement custom workflows (#1712) - as I see it, these can be modelled as a state machine, along with a permission policy that varies permissions on an object according to its current state.

kaedroho commented:
> Allow individual Wagtail installations to customise permission policies via settings (#1384). I haven't worked out the best way of doing this - could we pass it as an appconf parameter in INSTALLED_APPS, perhaps?

> Build up the group permission interface programmatically so that it outputs just the checkboxes relevant to the permission policy in use, rather than the app having to specifically register them. (For example - BlanketPermissionPolicy and AuthenticationOnlyPermissionPolicy would not add anything to the permissions interface, as there's nothing to configure; ModelPermissionPolicy would supply checkboxes for add, change and delete, and OwnershipPermissionPolicy would supply checkboxes for add and change, as it doesn't distinguish between change and delete permissions.)

I'm hoping that viewsets will provide a solution to both of these (#2113).

kaedroho commented:
Do you think it's worth having both ``BlanketPermissionPolicy`` and ``AuthenticationOnlyPermissionPolicy``? These basically mean the same thing as non-authenticated users will be turned away before they get to these model-level permissions anyway.

kaedroho commented:
Haven't checked yet if this is a pre-existing issue.

If I browse to a page that has an image chooser, but I do not have permissions to edit the selected image. I still get given the "Edit this image" button which just leads me to the dashboard with a big red banner saying "Sorry, you do not have permission to access this area."

kaedroho commented:
Thanks @gasman, this is really neat work!

Could you resolve: https://github.com/torchbox/wagtail/pull/2102#commitcomment-15504910

gasman commented:
> Do you think it's worth having both ``BlanketPermissionPolicy`` and ``AuthenticationOnlyPermissionPolicy``? These basically mean the same thing as non-authenticated users will be turned away before they get to these model-level permissions anyway.

The primary use-case for these policies is for sites that implement their own authorisation as a layer outside of Wagtail, as in #1384. I'd prefer not to make any assumptions about how these external systems will behave and what Wagtail needs to do to integrate correctly with them - for example, it might be done through a Django middleware that just sets `request.user` and does not do any permission enforcement itself (in which case Wagtail should reject anonymous users), or through a system like htaccess that doesn't touch Django at all (in which case we should allow them - making that work sensibly for wagtailadmin would require a non-trivial amount of additional work, of course). Since those implementations only take a few lines each, I figured it was worth leaving both options open.

I also want to leave open the possibility of these being used not just for admin interfaces - and indeed not just for Wagtail (there's nothing Wagtail-specific in the implementation). For something like a wiki, or a bug tracking tool, the distinction between the two policies is a lot more meaningful.

gasman commented:
Rebased and refactored to fix the issue with custom image/document models. I was wrong in my assessment that we couldn't define the permission_policy object statically at startup - we just can't do it in models.py (because the startup process specifically imports that).

Permission models are now defined in permissions.py, and since that implies having a wagtailcore/permissions.py to hold the permission policy for Site (it makes more sense to have it in the app that defines the model than in the app that edits it, I think), the actual implementation of the permission policies now lives in wagtailcore/permission_policies.py.

kaedroho commented:
Merged in 73b6ba01a551eb4065c8ecda6b343cff14487425 +parents. Thanks!
